homeassistant/components/tplink/entity.py

```diff
@@ -57,24 +79,201 @@ async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
+            _LOGGER.debug(
+                "Initializing feature-based %s with category %s",
+                self._attr_unique_id,
+                self._attr_entity_category,
+            )
+
+    def _category_for_feature(self, feature: Feature) -> EntityCategory | None:
+        """Return entity category for a feature."""
+        category_map = {
+            # Main controls have no category
+            Feature.Category.Primary: None,
+            Feature.Category.Config: EntityCategory.CONFIG,
+            Feature.Category.Info: EntityCategory.DIAGNOSTIC,
+            Feature.Category.Debug: EntityCategory.DIAGNOSTIC,
+        }
+        if (entity_category := category_map.get(feature.category)) is None:
+            _LOGGER.error(
+                "Unhandled category %s, fallback to DIAGNOSTIC", feature.category
+            )
+            entity_category = EntityCategory.DIAGNOSTIC
+
+        return entity_category
+
+    @abstractmethod
+    @callback
+    def _async_update_attrs(self) -> None:
+        """Platforms implement this to update the entity internals."""
+        raise NotImplementedError
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        try:
+            self._async_update_attrs()
+            self._attr_available = True
+        except Exception as ex:  # noqa: BLE001
+            if self._attr_available:
+                _LOGGER.warning(
+                    "Unable to read data for %s %s: %s",
+                    self.device,
+                    self.entity_description,
+                    ex,
+                )
+            self._attr_available = False
+
+        super()._handle_coordinator_update()
+
+
+def _entities_for_device[_E: CoordinatedTPLinkEntity](
+    device: Device,
+    *,
+    feature_type: Feature.Type,
+    entity_class: type[_E],
+    coordinator: TPLinkDataUpdateCoordinator,
+    parent: Device | None = None,
+) -> list[_E]:
+    """Return a list of entities to add.
+
+    This filters out unwanted features to avoid creating unnecessary entities
+    for device features that are implemented by specialized platforms like light.
+    """
+
+    def _filter(dev: Device, feat: Feature) -> bool:
+        if feat.type != feature_type:
+            return False
+
+        # We skip primary features for device types that have specialized platforms.
+        ignore_primary_controls_devicetypes = [
+            DeviceType.Bulb,
+            DeviceType.LightStrip,
+            DeviceType.Dimmer,
+            DeviceType.Thermostat,
+        ]
+        if (
+            feat.category == Feature.Category.Primary
+            and dev.device_type in ignore_primary_controls_devicetypes
+        ):
+            return False
+
+        return True
+
+    return [
+        entity_class(device, coordinator, feat, parent=parent)
+        for feat in device.features.values()
+        if _filter(device, feat)
```

Comment:
Yeah, that'd be faster but harder to read. I usually like to avoid too complex checks on comprehensions as it not only affects readability, but also makes it harder to set breakpoints for debugging.

I'm open for options to make this nicer :-)

Reply:
```python
    return [
        entity_class(device, coordinator, feat, parent=parent)
        for feat in device.features.values()
        if feat.type == feature_type
        and (
            feat.category != Feature.Category.Primary
            or device.device_type not in DEVICETYPES_WITH_SPECIALIZED_PLATFORMS
        )
    ]
```

Assuming this is what is intended

Reply:
Done, although I don't think it's worth for readability reasons.